hw/ip/otbn/dv/verilator/otbn_top_sim.cc

```diff
@@ -2,31 +2,110 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
+#include <cassert>
 #include <fstream>
+#include <getopt.h>
 #include <iomanip>
 #include <iostream>
+#include <memory>
+#include <sstream>
+#include <string>
 #include <svdpi.h>
+#include <vector>
 
+#include "log_trace_listener.h"
 #include "otbn_memutil.h"
+#include "otbn_trace_listener.h"
 #include "verilated_toplevel.h"
 #include "verilator_memutil.h"
 #include "verilator_sim_ctrl.h"
 
+std::vector<std::shared_ptr<OTBNTraceListener>> trace_listeners;
+
 extern "C" {
 extern unsigned int otbn_base_call_stack_get_size();
 extern unsigned int otbn_base_call_stack_get_element(int index);
 extern unsigned int otbn_base_reg_get(int index);
 extern unsigned int otbn_bignum_reg_get(int index, int quarter);
+
+extern void accept_otbn_trace_string(const char *trace,
+                                     unsigned int cycle_count) {
+  std::string trace_str(trace);
+
+  for (auto &l : trace_listeners) {
+    l->AcceptTraceString(trace, cycle_count);
+  }
+}
 }
 
+/**
+ * SimCtrlExtension that adds a '--trace_log_file' command line option. If set
+ * it sets up a LogTraceListener that will dump out the trace to the given log
+ * file
+ */
+class OTBNTraceUtil : public SimCtrlExtension {
+ private:
+  bool SetupTraceLog(const std::string &log_filename) {
+    try {
+      auto log_trace_listener =
+          std::make_shared<LogTraceListener>(log_filename);
+      trace_listeners.push_back(log_trace_listener);
+      return true;
+    } catch (const std::runtime_error &err) {
+      std::cerr << "ERROR: " << err.what() << std::endl;
```

Comment:
I would consider making this something like "ERROR: Failed to set up trace log: ...", which makes grepping much easier when weird stuff happens.